### PR TITLE
Pulse demon possession fixes

### LINKED ISSUE #35510: Pulse Demon Gun Hijack Sucks
(WEB REPORT BY: dovemite REMOTE: 209.222.101.3:7777)
# Revision
89f7b1d376c2349a70e8e5cd838e032e242ce3bd

# Description
See title, hijacking a gun sucks despite how it most likely would never occur normally. You cannot shoot the gun like a pAI slot gun could. Despite being in the gun, you'll die from lack of atmos. I don't think you can even get out of it if you leave the area where you do it
# Steps to Reproduce
Be Pulse Demon
Hijack a gun in a recharger

# What you Expected
Firing off lasers to cause mischief
Hitch hiking a ride
Being safe from atmos since I'm in a fucking gun
# What Actually Happened
Die from lack of pressure
No discernable way to leave the gun, can't wire hop or just walk out with the leave wire ability
No way to fire the gun

### LINKED ISSUE #35546: Pulse demon can't fire guns
(WEB REPORT BY: saulmyers REMOTE: 209.222.101.3:7777)
# Revision
dab5a4bce2c7791854ec45eacbd482b230dd11ed

# Description
Rechargers - Moves you inside the gun after a hijack delay, allowing you to fire it while inside, making people think the person wielding them is shooting them.
Tested with Spur and PKA, can't fire them.
# Steps to Reproduce
have someone place a gun into recharger, you are inside gun now.
# What you Expected
Behavior like smart laser gun and pAI
# What Actually Happened
Nothing.

### PR DESCRIPTION
**Pull request by SECBATON-GRIFFON**
[bugfix][tweak]

## What this does
creates an action button that allows you to return to an APC when inside something possessed
makes toggle drain into one too from a spell
makes gun firing work (Closes #35510. Closes #35546.)

## Why it's good
easier escape of possessed items

## Changelog
:cl:
 * rscadd: Pulse demons can now exit items they are possessing with an action button on their HUD
 * rscadd: Adds some backend stuff for admins to test pulse demons with, namely allowing them to instantly possess guns and robots without the intermediary means, and a spawn-only type having everything maxed and unlocked
 * bugfix: Pulse demons can now fire guns they possess properly
 * tweak: Pulse demons now no longer need to be in their original APC area to override robot targeting systems
 * tweak: Toggle drain is now an action button, not a spell

### CODE REVIEW

File: code/modules/mob/living/simple_animal/hostile/pulse_demon/pulsedemon_powers.dm [hunk 1]
```diff
@@ -630,3 +608,35 @@
...
 	..()
+
+/datum/action/pd_toggle_drain
+	name = "Toggle power drain"
+	desc = "Toggles the draining of power while in an APC, battery or cable"
+	icon_icon = 'icons/mob/screen_spells.dmi'
+	button_icon_state = "pd_toggle"
+
+/datum/action/pd_toggle_drain/Trigger()
+	if(istype(owner,/mob/living/simple_animal/hostile/pulse_demon))
```
Comment (jwhitak) [inline code/modules/mob/living/simple_animal/hostile/pulse_demon/pulsedemon_powers.dm]: This was probably written before the ispulsedemon define huh?

Comment (SECBATON-GRIFFON) [reply code/modules/mob/living/simple_animal/hostile/pulse_demon/pulsedemon_powers.dm]: not sure, but fixed

**Comment by boy2mantwicethefam** [comment]
Pulse Demons also need to go back to the original APC if they're possessing a bot.

**Comment by SECBATON-GRIFFON** [comment]
> Pulse Demons also need to go back to the original APC if they're possessing a bot.

i added the button that lets them do that for bots too

File: code/modules/mob/living/simple_animal/hostile/pulse_demon/pulsedemon.dm [hunk 2]
```diff
@@ -78,16 +79,36 @@
...
+	health_regen_rate = INFINITY
+	amount_per_regen = 1
+	charge_absorb_amount = INFINITY
+//	max_can_absorb = INFINITY
+	takeover_time = 1
+	move_divide = 1
+
+/mob/living/simple_animal/hostile/pulse_demon/maxedout/New()
+	..()
+	for(var/spell/S in possible_spells)
```
Comment (west3436) [inline code/modules/mob/living/simple_animal/hostile/pulse_demon/pulsedemon.dm:107]: you can use pick_n_take here

**Review by Kurfursten — CHANGES_REQUESTED** [review]
implement suggestion

File: code/modules/mob/living/simple_animal/hostile/pulse_demon/pulsedemon.dm (hunk 2 above)
Comment (SECBATON-GRIFFON) [reply code/modules/mob/living/simple_animal/hostile/pulse_demon/pulsedemon.dm:107]: howso?

Comment (Kurfursten) [reply code/modules/mob/living/simple_animal/hostile/pulse_demon/pulsedemon.dm:107]:
```
var/spell/S = pick_n_take(possible_spells)
add_spell(S,otherargs)
```

Comment (SECBATON-GRIFFON) [reply code/modules/mob/living/simple_animal/hostile/pulse_demon/pulsedemon.dm:107]: won't that add them in a random order though?

Comment (Kurfursten) [reply code/modules/mob/living/simple_animal/hostile/pulse_demon/pulsedemon.dm:107]: You pose a good point, if you care about the order you really shouldn't use that helper here which is explicitly trying to randomize

**Review by Kurfursten — APPROVED** [review]

File: code/modules/mob/living/simple_animal/hostile/pulse_demon/pulsedemon.dm (hunk 2 above)
Comment (SECBATON-GRIFFON) [reply code/modules/mob/living/simple_animal/hostile/pulse_demon/pulsedemon.dm:107]: i'd rather not mess with a tester's muscle memory so order is a nicer thing to have

**Pull request merged**